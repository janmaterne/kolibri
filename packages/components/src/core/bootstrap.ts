import type { Generic, LoaderCallback, RegisterOptions } from 'adopted-style-sheets';
import { register as coreRegister } from 'adopted-style-sheets';
import { setDevMode } from '../schema';
import { setCustomTagNames } from './component-names';
import { initializeI18n } from './i18n';

type Environment = 'development' | 'production';

type KoliBriOptions = RegisterOptions & {
	/**
	 * The environment in which the application is running.
	 */
	environment?: Environment;
	/**
	 * This option allows you to transform the component tag names.
	 */
	transformTagName?: (tagName: string) => string;
};

let initialized = false;

export const bootstrap = async (
	themes:
		| Generic.Theming.RegisterPatch<string, string, string>
		| Generic.Theming.RegisterPatch<string, string, string>[]
		| Set<Generic.Theming.RegisterPatch<string, string, string>>,
	loaders: LoaderCallback | LoaderCallback[] | Set<LoaderCallback>,
	options?: KoliBriOptions,
): Promise<void[]> => {
	setDevMode(options?.environment === 'development');

	initializeI18n(options?.translation?.name ?? 'de', options?.translations);
	if (options?.transformTagName) {
		setCustomTagNames(options?.transformTagName);
	}
	const coreRegisterReturnValue = await coreRegister(themes, loaders, options);
	initialized = true;

	return coreRegisterReturnValue;
};

export const register = bootstrap;
export const isInitialized = () => initialized;
