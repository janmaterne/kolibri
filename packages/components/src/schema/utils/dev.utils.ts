/* eslint-disable no-console */

type GetModeFn = () => boolean;

let WINDOW: Window | null = null;
let DOCUMENT: Document | null = null;

export const getWindow = (): Window => (WINDOW || typeof window === 'undefined' ? (null as unknown as Window) : window);
export const setWindow = (value: Window): void => {
	WINDOW = value;
};

export const getDocument = (): Document => (DOCUMENT || typeof getWindow().document === 'undefined' ? (null as unknown as Document) : getWindow().document);
export const setDocument = (value: Document): void => {
	DOCUMENT = value;
};

let DEV_MODE: boolean = false;
let EXPERIMENTAL_MODE: boolean = false;
let COLOR_CONTRAST_ANALYSIS: boolean = false;

export const getDevMode = (): boolean => DEV_MODE === true;
export const setDevMode = (mode: boolean): void => {
	DEV_MODE = mode === true;
};

export const getExperimentalMode = (): boolean => EXPERIMENTAL_MODE === true;
export const setExperimentalMode = (mode: boolean): void => {
	EXPERIMENTAL_MODE = mode === true;
};

export const getColorContrastAnalysis = (): boolean => COLOR_CONTRAST_ANALYSIS === true;
export const setColorContrastAnalysis = (mode: boolean): void => {
	COLOR_CONTRAST_ANALYSIS = mode === true;
};

type LogShieldOptions = {
	classifier?: string;
	forceLog?: boolean;
	overwriteStyle?: string;
};

const LOG_STYLE = 'color: white; background: #666; font-weight: bold; padding: .25em .5em; border-radius: 3px; border: 1px solid #000';

const mapToArray = (msg: unknown | unknown[]): unknown[] => {
	return Array.isArray(msg) ? msg : [msg];
};

const getLogLabel = (label: string): string => {
	return `%c${label}`;
};

const handleClassifier = (label: string, classifier?: string): string => {
	if (typeof classifier === 'string' && classifier.length > 0) {
		return `${getLogLabel(label)} | ${classifier}`;
	} else {
		return getLogLabel(label);
	}
};

const getShield = (label: string, options?: LogShieldOptions): string[] => {
	return [handleClassifier(label, options?.classifier), `${LOG_STYLE};${options?.overwriteStyle || ''}`];
};

const isDevModeOrForceLog = (devMode: GetModeFn, forceLog?: boolean): boolean => devMode() || forceLog === true;

export class Logger {
	public constructor(
		private readonly label: string,
		private readonly devMode: GetModeFn,
	) {}

	public debug(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (isDevModeOrForceLog(this.devMode, options?.forceLog)) {
			console.debug(...getShield(this.label, options), ...mapToArray(msg));
		}
	}

	public info(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (isDevModeOrForceLog(this.devMode, options?.forceLog)) {
			console.info(...getShield(this.label, options), ...mapToArray(msg));
		}
	}

	public trace(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (isDevModeOrForceLog(this.devMode, options?.forceLog)) {
			console.trace(...getShield(this.label, options), ...mapToArray(msg));
		}
	}

	public warn(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (isDevModeOrForceLog(this.devMode, options?.forceLog)) {
			console.warn(...getShield(this.label, options), ...mapToArray(msg));
		}
	}

	public error(msg: unknown | unknown[], options?: LogShieldOptions): void {
		console.error(...getShield(this.label, options), ...mapToArray(msg));
	}

	public throw(msg: unknown | unknown[], options?: LogShieldOptions): void {
		if (isDevModeOrForceLog(this.devMode, options?.forceLog)) {
			throw new Error(...getShield(this.label, options), ...mapToArray(msg));
		}
	}
}

export const Log = new Logger('KoliBri', getDevMode);
