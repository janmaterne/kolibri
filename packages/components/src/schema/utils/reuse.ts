const PROCESS_ENVS = ['development', 'production', 'test'] as const;
type ProcessEnv = (typeof PROCESS_ENVS)[number];
export let processEnv: ProcessEnv = 'development';
try {
	processEnv = process.env.NODE_ENV as ProcessEnv;
} catch (e) {
	processEnv = 'production';
}

/**
 * This function is used to handle the slot content by
 * moving a DOM element to a defined slot element.
 */
export const handleSlotContent = (hostRef: HTMLElement, slotRef: HTMLElement, slotName: string) => {
	if (hostRef && slotRef && typeof slotName === 'string') {
		const content = hostRef.querySelector(`[slot="${slotName}"]`);
		if (content) {
			slotRef.appendChild(content);
		}
	}
};

/**
 * This function is used to check if a label is an empty string.
 * @param label The label of the slot
 * @returns  True if the label is an empty string
 */
export const showExpertSlot = (label: unknown) => label === '';

/**
 * This function builds the BadgeText. It chooses access key over shortkey or empty string, if both are empty
 * @param accessKey
 * @param shortKey
 * @returns returns badgeText as string
 */
export const buildBadgeTextString = (accessKey?: string, shortKey?: string) => {
	return accessKey || shortKey || '';
};
