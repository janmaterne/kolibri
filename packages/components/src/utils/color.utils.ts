import colorConvert from 'color-convert';
import { CMYK } from 'color-convert/conversions';
import { hex as hexContrast, score } from 'wcag-contrast';

export const cmykSaturation = (cmyk: number[], saturation = 1): CMYK => {
	return [cmyk[0] * saturation, cmyk[1] * saturation, cmyk[2] * saturation, cmyk[3] * saturation];
};

export const getHexBySat = (hex: string, saturation = 1): string => {
	return colorConvert.cmyk.hex(cmykSaturation(colorConvert.hex.cmyk(hex), saturation));
};

interface KoliBriContrast {
	color: {
		background: string;
		text: string;
	};
	contrast: {
		value: number;
		score: string;
	};
}

export const getValues = (hex: string, saturation = 1): KoliBriContrast => {
	const bgColor = getHexBySat(hex, saturation);
	let fgColor = null;
	let bgFgScore = null;
	let bgFgValue = null;
	const black = hexContrast(bgColor, '000');
	const white = hexContrast(bgColor, 'fff');
	if (black >= white) {
		fgColor = '000';
		bgFgScore = score(black);
		bgFgValue = black;
	} else {
		fgColor = 'fff';
		bgFgScore = score(white);
		bgFgValue = white;
	}
	return {
		color: {
			background: bgColor,
			text: fgColor,
		},
		contrast: {
			value: bgFgValue,
			score: bgFgScore,
		},
	};
};
