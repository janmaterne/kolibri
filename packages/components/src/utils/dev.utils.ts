import { Log, getDocument, processEnv, setColorContrastAnalysis, setDevMode, setExperimentalMode } from '../schema';

import { getWindow } from '../schema';

const initMeta = (): void => {
	const meta = getDocument().querySelector('meta[name="kolibri"]');
	if (meta && meta.hasAttribute('content')) {
		const content = meta.getAttribute('content');
		if (typeof content === 'string') {
			setDevMode(content.includes('dev-mode=true'));
			setExperimentalMode(content.includes('experimental-mode=true'));
			setColorContrastAnalysis(content.includes('color-contrast-analysis=true'));
		}
	}
};

const getKoliBri = (): Record<string, unknown> => {
	let kolibri = getWindow().KoliBri;
	if (kolibri === undefined) {
		kolibri = {};
		Object.defineProperty(getWindow(), 'KoliBri', {
			value: kolibri,
			writable: false,
		});
	}
	return kolibri;
};

export { getKoliBri };

export const initKoliBri = (): void => {
	initMeta();
	Log.debug(
		`
,--. ,--.         ,--. ,--. ,-----.           ,--.
|  .'   /  ,---.  |  | \`--' |  |) /_  ,--.--. \`--'
|  .   '  | .-. | |  | ,--. |  .-.  \\ |  .--' ,--.
|  |\\   \\ | '-' | |  | |  | |  '--' / |  |    |  |
\`--' \`--Â´  \`---Â´  \`--' \`--' \`------Â´  \`--'    \`--'
ðŸš¹ The accessible HTML-Standard | ðŸ‘‰ https://public-ui.github.io | 2.1.1
	`,
		{
			forceLog: true,
		},
	);
};

export const renderDevAdvice = (): void => {
	if (getKoliBri().adviceShown !== true) {
		Object.defineProperty(getKoliBri(), 'adviceShown', {
			get: function () {
				return true;
			},
		});
		Log.debug(
			`
You are using the KoliBri component library. If you have any suggestions for improvement or find a problem, please contact us:

Ticket: https://github.com/public-ui/kolibri/issues/new/choose (for privacy reasons, please use email)
Email: kolibri@itzbund.de
`,
		);
	}
};

let nonce = (): string => Math.floor(Math.random() * 16777215).toString(16);

if (processEnv === 'test') {
	nonce = (): string => 'nonce';
}

export { nonce };
