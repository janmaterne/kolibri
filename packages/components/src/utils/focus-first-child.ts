// Types for the configuration options
interface FocusOptions {
	button?: boolean;
	link?: boolean;
	textarea?: boolean;
	inputText?: boolean;
	inputRadio?: boolean;
	inputCheckbox?: boolean;
	inputNumber?: boolean;
	inputEmail?: boolean;
	inputTel?: boolean;
	inputPassword?: boolean;
	inputDate?: boolean;
	inputTime?: boolean;
	inputFile?: boolean;
	select?: boolean;
	tabindex?: boolean;
}

/**
 * Focus the first focusable child element within a parent element
 * @param element - The parent element to search within
 * @param options - Configuration object to specify which elements should be focusable
 * @returns boolean - True if an element was focused, false otherwise
 */
export function focusFirstChild(element: HTMLElement, options: FocusOptions = {}): boolean {
	// Default options configuration
	const defaultOptions: Required<FocusOptions> = {
		button: true,
		link: true,
		textarea: true,
		inputText: true,
		inputRadio: true,
		inputCheckbox: true,
		inputNumber: true,
		inputEmail: true,
		inputTel: true,
		inputPassword: true,
		inputDate: true,
		inputTime: true,
		inputFile: true,
		select: true,
		tabindex: true,
	};

	// Merge default options with provided options
	const settings: Required<FocusOptions> = { ...defaultOptions, ...options };

	// Build selectors array based on enabled options
	const selectors: string[] = [];

	if (settings.button) selectors.push('button:not([disabled])');
	if (settings.link) selectors.push('a[href]:not([disabled])');
	if (settings.textarea) selectors.push('textarea:not([disabled])');
	if (settings.select) selectors.push('select:not([disabled])');
	if (settings.tabindex) selectors.push('[tabindex]:not([tabindex="-1"]):not([disabled])');

	// Configure input types based on settings
	const inputTypes: string[] = [];
	if (settings.inputText) inputTypes.push('text');
	if (settings.inputRadio) inputTypes.push('radio');
	if (settings.inputCheckbox) inputTypes.push('checkbox');
	if (settings.inputNumber) inputTypes.push('number');
	if (settings.inputEmail) inputTypes.push('email');
	if (settings.inputTel) inputTypes.push('tel');
	if (settings.inputPassword) inputTypes.push('password');
	if (settings.inputDate) inputTypes.push('date');
	if (settings.inputTime) inputTypes.push('time');
	if (settings.inputFile) inputTypes.push('file');

	// Add input selectors if any input types are enabled
	if (inputTypes.length > 0) {
		selectors.push(inputTypes.map((type) => `input[type="${type}"]:not([disabled])`).join(', '));
	}

	// Combine all selectors
	const selectorString = selectors.join(', ');

	// Return false if no selectors are available (all options disabled)
	if (!selectorString) {
		return false;
	}

	// Query for all focusable elements
	const focusableElements = element.querySelectorAll<HTMLElement>(selectorString);

	// Find first visible and focusable element
	const firstFocusable = Array.from(focusableElements).find((el) => {
		return !!(el.offsetWidth || el.offsetHeight || el.getClientRects().length);
	});

	if (firstFocusable) {
		firstFocusable.focus();
		return true;
	}
	return false;
}
