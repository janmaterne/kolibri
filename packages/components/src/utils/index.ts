export * from './isDefined';
export * from './isObject';
export * from './isString';
export * from './isStringOrEmpty';

export * from './markdown';
