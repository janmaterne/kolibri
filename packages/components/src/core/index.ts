import type { Generic, LoaderCallback, RegisterOptions } from '@a11y-ui/core';
import { register as coreRegister } from '@a11y-ui/core';

import { I18nextService, II18nService } from './i18n';
import { KoliBri } from '../utils/dev.utils';

export const register = async (
	themes:
		| Generic.Theming.RegisterPatch<string, string, string>
		| Generic.Theming.RegisterPatch<string, string, string>[]
		| Set<Generic.Theming.RegisterPatch<string, string, string>>,
	loaders: LoaderCallback | LoaderCallback[] | Set<LoaderCallback>,
	options?: RegisterOptions
): Promise<void[]> => {
	if (KoliBri.I18n === undefined) {
		Object.defineProperty(KoliBri, 'I18n', {
			value: await I18nextService.createInstance(options?.translation?.name ?? 'de', options?.translations),
			writable: false,
		});
	}
	return await coreRegister(themes, loaders, options);
};

export const getI18nService: () => II18nService | undefined = () => KoliBri.I18n as II18nService;
