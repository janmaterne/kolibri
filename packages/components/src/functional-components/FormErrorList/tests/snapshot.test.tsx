import { h } from '@stencil/core';
import FormErrorListFc from '../FormErrorList';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';
import * as i18n from '../../../i18n';

describe('FormErrorListFc', () => {
	beforeEach(() => {
		// Mock the translate function directly on the i18n object
		jest.spyOn(i18n, 'translate').mockImplementation((key: string) => {
			switch (key) {
				case 'kol-error-list-message':
					return 'Fehlerübersicht';
				case 'kol-error-list':
					return 'Liste der Fehler';
				default:
					return key;
			}
		});
	});

	afterEach(() => {
		jest.restoreAllMocks();
	});

	it('should render null when errorList is empty', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <FormErrorListFc errorList={[]} />);
		expect(page.root).toBeNull();
	});

	it('should render with error list and translated label', async () => {
		const errorList = [
			{
				message: 'Fehler 1',
				selector: '#field1',
			},
		];

		const page = await renderFunctionalComponentToSpecPage(() => <FormErrorListFc errorList={errorList} />);

		expect(page.root).toMatchSnapshot();
		const headingElement = page.root?.querySelector('kol-heading-wc');
		expect(headingElement?.getAttribute('_label')).toBe('Fehlerübersicht');

		const navElement = page.root?.querySelector('nav');
		expect(navElement?.getAttribute('aria-label')).toBe('Liste der Fehler');

		const linkElement = page.root?.querySelector('kol-link-wc');
		expect(linkElement).toBeTruthy();
		expect(linkElement?.getAttribute('_label')).toBe('Fehler 1');
	});
});
