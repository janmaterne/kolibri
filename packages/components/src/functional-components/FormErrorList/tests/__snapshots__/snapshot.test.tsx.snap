// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormErrorListFc should render with error list and translated label 1`] = `
<div class="alert card error kol-alert-wc">
  <div class="heading">
    <kol-icon _icons="codicon codicon-error" _label="kol-error" class="heading-icon"></kol-icon>
    <div class="heading-content">
      <kol-heading-wc _label="FehlerÃ¼bersicht"></kol-heading-wc>
    </div>
  </div>
  <div class="content">
    <nav aria-label="Liste der Fehler">
      <ul>
        <li>
          <kol-link-wc _href="#field1" _label="Fehler 1"></kol-link-wc>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
